### PR TITLE
ON-45096 # added MFA components

### PR DESCRIPTION
**Pull request by aaronroworth**
## Requester Checklist

Please only check the items that you have actioned. Do not check items that are not applicable to your PR.

### Implementation

- [x] Have you tested your implementation locally?
- [x] If applicable, has an appropriate changelog entry been added? Do not include if you are fixing/changing something that is only in the current release.
- [ ] There are no warnings that have been suppressed unnecessarily
- [ ] Have automated tests been added, or have related ones been updated to cover the change?
- [ ] Have all OneBlink dependency updates been completed (eg apps / apps-react / types etc).
- [ ] Have you ensured this change does not add unwanted dependencies?
- [ ] If this PR contains a refactor, have relevant Jira testing tasks added?
- [ ] Changes that will knowingly make the feature/bug incomplete have been commented with TODO and a description of what needs to be done to finish the feature/bug
- [ ] Any changes made to public APIs have been reflected in the documentation
- [ ] Have you isolated business logic where possible to allow for unit testing?

### Logging and Debugging

- [ ] Are the error messages, if any, informative?
- [ ] Are there enough log events and are they written in a way that allows for easy debugging?
- [x] "Debugging" code removed
- [x] Front-end: No erroneous `Console.WriteLines`

### Readability

- [ ] All class, variable, property and method modifiers are provided with the smallest scope possible
- [x] New files, variables and functions are descriptive/comprehensible and named consistently.
- [x] There is no dead code (unreachable code)
- [x] There is no usage of [magic numbers](<https://en.wikipedia.org/wiki/Magic_number_(programming)>)
- [x] There is no commented out code.
- [ ] In hard-to-understand areas, comments exist and describe rationale or reasons for decisions in code

### Security

- [ ] All personal data inputs are checked (for the correct type, length/size, format, and range).
- [ ] No sensitive information is logged or visible in a stacktrace
- [ ] Are authorization and authentication handled correctly?
- [ ] Is (user) input validated, sanitized, and escaped to prevent security attacks such as cross-site scripting or SQL injection?
- [ ] Is data retrieved from external APIs or libraries checked for security issues?
- [ ] Do API endpoints return appropriate status codes

## Reviewer Guide

- It is important that you understand the purpose of the PR.
- You are encouraged to engage with the requestor if you do not understand any of the proposed code changes/additions/deletions.
- Do you, the reviewer, understand what the code does? Do you think a specific expert, like a security expert or a usability expert, should look over the code before it can be accepted?
- Is a framework, API, library, or service used that should not be used? Are there alternatives you could recommend?
- Are there existing hooks/components/functions in the same code base that could be utilised?
- When reviewing tests, attempt to identify missing edge cases that may be relevant to the proposed implementation
- Ensure you check for:
  - Security
  - Scalability
  - Performance
  - Maintainability

### CODE REVIEW

**Review by kizaonline — APPROVED** [review]

File: src/components/mfa/MultiFactorAuthentication.tsx [hunk 1]
```diff
@@ -0,0 +1,244 @@
...
+                      }
+                    />
+                  )}
+                </Typography>
+                <Box marginY={1}>
+                  <Divider />
+                </Box>
+                <Typography variant="body2" paragraph>
+                  Multi factor authentication (MFA), also known as two factor
+                  authentication (2FA), is a best practice that requires a
```
Comment (RyanButton) [inline src/components/mfa/MultiFactorAuthentication.tsx:139]: not quite true ?

https://www.incognia.com/the-authentication-reference/what-are-the-key-differences-between-2fa-and-mfa#:~:text=MFA%20vs%202FA,all%20MFA%20is%20a%202FA.

File: src/components/mfa/MultiFactorAuthentication.tsx [hunk 2]
```diff
@@ -0,0 +1,244 @@
...
+  return (
+    <Grid item xs={true} lg={8}>
+      <Box padding={3}>
+        <Paper>
+          <Box padding={3}>
+            <Grid container spacing={2} alignItems="center">
+              <Grid item xs>
+                <Typography variant="h4" fontWeight="light">
+                  Multi Factor Authentication{' '}
+                  {isExternalIdentityProviderUser ? undefined : isLoading ? (
```
Comment (RyanButton) [inline src/components/mfa/MultiFactorAuthentication.tsx:99]: Would be good if we could avoid the long ternaries, maybe rip this out into a new sub-component and render it like :

if(isExternalIdentityProviderUser) {
   return undefined
}
if(isLoading) {
  return loading component
}

etc

File: src/components/mfa/MultiFactorAuthentication.tsx [hunk 3]
```diff
@@ -0,0 +1,244 @@
...
+                </Box>
+                <Typography variant="body2" paragraph>
+                  Multi factor authentication (MFA), also known as two factor
+                  authentication (2FA), is a best practice that requires a
+                  second authentication factor in addition to user name and
+                  password sign-in credentials. We strongly recommend enabling
+                  MFA to enhance your account security.
+                </Typography>
+                <Grid container justifyContent="flex-end" spacing={1}>
+                  {ssoSetupUrl ? (
```
Comment (RyanButton) [inline src/components/mfa/MultiFactorAuthentication.tsx:145]: again could probably put this into a sub component to make it more readable

File: src/hooks/useAppUserMfa.tsx [hunk 4]
```diff
@@ -0,0 +1,282 @@
...
+ * @param props
+ * @returns
+ * @group Components
+ */
+export function AppUserMfaProvider({
+  children,
+}: {
+  children: React.ReactNode
+}) {
+  const userProfile = authService.getUserProfile()
```
Comment (RyanButton) [inline src/hooks/useAppUserMfa.tsx:85]: will this work in the console?
